accord-core/src/main/java/accord/local/CommandStore.java

```diff
@@ -673,15 +808,45 @@ protected void updatedRedundantBefore(SafeCommandStore safeStore, RedundantBefor
         listeners.clearBefore(this, clearWaitingBefore);
     }
 
-    protected final Ranges isWaitingOnSync(TxnId syncId, Ranges ranges)
+    public AsyncResult<Void> awaitVisibility(long epoch, Ranges ranges)
+    {
+        synchronized (waitingOnVisibility)
+        {
+            if (waitingOnVisibility.isEmpty())
+                return AsyncResults.success(null);
+
+            List<AsyncResult<Void>> awaiting = new ArrayList<>();
+            Ranges waitingOn = Ranges.EMPTY;
+            for (Map.Entry<Long, WaitingOnVisibility> e : waitingOnVisibility.entrySet())
+            {
+                if (e.getKey() > epoch)
+                    break;
+
+                Ranges remaining = e.getValue().waitingOn;
+                Ranges intersecting = remaining.slice(ranges, Minimal);
+                if (!intersecting.isEmpty())
+                {
+                    awaiting.add(e.getValue().whenDone);
+                    ranges = ranges.without(intersecting);
+                    waitingOn = waitingOn.with(intersecting);
```

Comment:
Do we need to accumulate it here if we never use it?

Reply:
nope!